cartoframes/data/services/geocode.py

```diff
@@ -0,0 +1,472 @@
+
+
+def _generate_temp_table_name(base=None):
+    return (base or 'table') + '_' + uuid.uuid4().hex[:10]
+
+
+def _set_pre_summary_info(summary, output):
+    logging.debug(summary)
+    output['total_rows'] = sum(summary.values())
+    # TODO: either report new (ng+nn) and changed (cg+cn) records, or we could
+    # reduce the states by merging ng+cg, nn+nn
+    output['required_quota'] = sum([summary[s] for s in ['ng', 'nn', 'cg', 'cn']])
+    output['previously_geocoded'] = summary.get('pg', 0)
+    output['previously_failed'] = summary.get('pn', 0)
+    output['records_with_geometry'] = sum([summary[s] for s in ['ng', 'cg', 'pg']])
+    # output['records_without_geometry'] = sum([summary[s] for s in ['nn', 'cn', 'pn']])
+
+
+def _set_post_summary_info(summary, result, output):
+    if result and result.get('total_rows', 0) == 1:
+        null_geom_count = result.get('rows')[0].get('count')
+        geom_count = output['total_rows'] - null_geom_count
+        output['final_records_with_geometry'] = geom_count
+        # output['final_records_without_geometry'] = null_geom_count
+        output['geocoded_increment'] = output['final_records_with_geometry'] - output['records_with_geometry']
+        output['successfully_geocoded'] = output['geocoded_increment'] + sum([summary[s] for s in ['ng', 'cg']])
+        output['failed_geocodings'] = output['required_quota'] - output['successfully_geocoded']
+
+
+class Geocode(object):
+    """Geocode using CARTO data services.
+    This requires a CARTO account; master API Key credentials must be provided
+    (throuch explicit argument in contructor or via the default credentials)
+    to access the service. Use of these methods will incurn in geocoding
+    credit consumption for the privided accout.
+
+    Examples:
+
+        Obtain the number of credits needed to geocode a dataset:
+
+        .. code::
+
+            from data.services import Geocode
+            gc = Geocode(credentials)
+            _, info = gc.geocode(dataset, street='address', dry_run=True)
+            print(info['required_quota'])
+
+        Geocode a dataframe:
+
+        .. code::
+
+            from data.services import Geocode
+            from cartoframes.data import Dataset
+            import pandas
+
+            dataframe = pandas.DataFrame([['Gran Vía 46', 'Madrid'], ['Ebro 1', 'Sevilla']], columns=['address','city'])
+            gc = Geocode()
+            geocoded_dataset, info = gc.geocode(dataframe, street='address', city='city', country="'Spain'")
+            print(dataframe)
+
+        Geocode a table:
+
+        .. code::
+
+            from data.services import Geocode
+            from cartoframes.data import Dataset
+            import pandas
+
+            dataframe = pandas.DataFrame([['Gran Vía 46', 'Madrid'], ['Ebro 1', 'Sevilla']], columns=['address','city'])
+            dataset = Dataset(dataframe)
+            dataset.upload(table_name='offices')
```

Comment:
I don't think we should try to upload anything on behalf of the user (we do it internally into temporary tables only). But I think the example is unfortunate: I tried to demonstrate geocoding a table, but since I'm using a dataframe to create the table first it gets quite confusing (and we could have geocoded the dataframe directly and save results to a table).

Reply:
In the upload you will need a credentials instance or use `set_default_credentials` before in the same way as you did in the documentation

About the upload, after showing the Geocode code, I think uploading it before geocoding makes no sense as you said before